### PR TITLE
[deckhouse] improve module loader

### PR DESCRIPTION
**Pull request by ipaqsa**
## Description
It provides: 
- refactor for the module release controller.
- refactor for the module override controller(make it separate).
- improvement for the module loader(sync fs and cluster at start).

## Why do we need it, and what problem does it solve?
The module release and override controller should not do sync tasks. We have the module loader, and it has to sync FS and cluster before controllers start. 

## What is the expected result?

Controllers inform that they are ready, it happens after modules inited and controllers preflight done

Screenshot 2024-11-28 at 1 46 08 AM

Try to remove module from FS:

1. Remove module from FS
```
[deckhouse] deckhouse@dev-master-0 /deckhouse $ rm -rf downloaded/modules/901-test
```

3. Restart deckhouse

4. Module loader restored the 'test' module:

```
{"level":"debug","logger":"deckhouse-controller.module-loader","msg":"init module loader","source":"deckhouse/deckhouse-controller/pkg/controller/moduleloader/loader.go:101","time":"2024-11-27T22:56:45Z"}
{"level":"debug","logger":"deckhouse-controller.module-loader","msg":"restore absent modules from overrides","source":"deckhouse/deckhouse-controller/pkg/controller/moduleloader/loader.go:103","time":"2024-11-27T22:56:45Z"}
{"level":"debug","logger":"deckhouse-controller.module-loader","msg":"restore absent modules from releases","source":"deckhouse/deckhouse-controller/pkg/controller/moduleloader/loader.go:108","time":"2024-11-27T22:56:45Z"}
{"level":"info","logger":"deckhouse-controller.module-loader","msg":"the 'test' module symlink is absent on file system, restore it","source":"deckhouse/deckhouse-controller/pkg/controller/moduleloader/sync.go:170","time":"2024-11-27T22:56:45Z"}
{"level":"info","logger":"deckhouse-controller.module-loader","msg":"the 'test' module is absent on filesystem, restore it from the 'test' source","source":"deckhouse/deckhouse-controller/pkg/controller/moduleloader/sync.go:254","time":"2024-11-27T22:56:45Z"}
{"level":"info","logger":"deckhouse-controller.module-loader","msg":"the 'test/v0.8.5' module restored to ../test/v0.8.5","source":"deckhouse/deckhouse-controller/pkg/controller/moduleloader/sync.go:278","time":"2024-11-27T22:56:45Z"}
```

5. Check fs
```
[deckhouse] deckhouse@dev-master-0 /deckhouse $ ls downloaded/modules/
901-test
```

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse
type: feature
summary: Improve module loader.
impact_level: low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @ipaqsa. I've started the workflow for label 'e2e/run/gcp'...

File: deckhouse-controller/pkg/controller/module-controllers/config/controller.go [hunk 1]
```diff
@@ -339,11 +352,16 @@ func (r *reconciler) deleteModuleConfig(ctx context.Context, moduleConfig *v1alp
...
 			controllerutil.RemoveFinalizer(moduleConfig, v1alpha1.ModuleConfigFinalizer)
-			return true
+			needsUpdate = true
 		}
-		return false
+		if _, ok := moduleConfig.ObjectMeta.Annotations[v1alpha1.ModuleConfigAnnotationAllowDisable]; ok {
+			delete(moduleConfig.ObjectMeta.Annotations, v1alpha1.ModuleConfigAnnotationAllowDisable)
+			needsUpdate = true
+		}
+		return needsUpdate
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/config/controller.go:364]:
```suggestion
		if controllerutil.ContainsFinalizer(moduleConfig, v1alpha1.ModuleConfigFinalizer) {
			controllerutil.RemoveFinalizer(moduleConfig, v1alpha1.ModuleConfigFinalizer)
			return true
		}
		if _, ok := moduleConfig.ObjectMeta.Annotations[v1alpha1.ModuleConfigAnnotationAllowDisable]; ok {
			delete(moduleConfig.ObjectMeta.Annotations, v1alpha1.ModuleConfigAnnotationAllowDisable)
			return true
		}
        
		return false
```

File: deckhouse-controller/pkg/controller/module-controllers/config/sync.go [hunk 2]
```diff
@@ -38,17 +37,14 @@ func (r *reconciler) syncModules(ctx context.Context) error {
...
-	for _, moduleName := range r.moduleManager.GetModuleNames() {
-		module := new(v1alpha1.Module)
-		if err := r.client.Get(ctx, client.ObjectKey{Name: moduleName}, module); err != nil {
-			if apierrors.IsNotFound(err) {
-				r.log.Warnf("the '%s' module not found", moduleName)
-				continue
-			}
-			return fmt.Errorf("get the '%s' module: %w", moduleName, err)
-		}
+	r.log.Debugf("sync modules")
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/config/sync.go:40]:
```suggestion
	r.log.Debug("sync modules")
```

File: deckhouse-controller/pkg/controller/module-controllers/config/sync.go [hunk 3]
```diff
@@ -76,58 +72,21 @@ func (r *reconciler) syncModules(ctx context.Context) error {
...
 	}
-	r.log.Debug("registered modules are inited, init module configs")
-
-	if err := r.syncModuleConfigs(ctx); err != nil {
-		return fmt.Errorf("sync module configs: %w", err)
-	}
-
-	r.log.Debug("module configs are inited, run event loop")
 
+	r.log.Debugf("controller is ready")
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/config/sync.go:86]:
```suggestion
	r.log.Debug("controller is ready")
```

File: deckhouse-controller/pkg/controller/moduleloader/loader.go [hunk 4]
```diff
@@ -69,20 +72,55 @@ type Loader struct {
...
+	if err := l.restoreAbsentModulesFromReleases(ctx); err != nil {
+		return fmt.Errorf("restore absent modules from releases: %w", err)
+	}
+
+	l.log.Debugf("delete modules with absent release")
+	if err := l.deleteModulesWithAbsentRelease(ctx); err != nil {
+		return fmt.Errorf("delete modules with absent releases: %w", err)
+	}
+
+	l.log.Debugf("module loader initialized")
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/moduleloader/loader.go:119]:
```suggestion
	l.log.Debug("module loader initialized")
```

File: deckhouse-controller/pkg/controller/module-controllers/source/controller.go [hunk 5]
```diff
@@ -152,6 +152,8 @@ func (r *reconciler) preflight(ctx context.Context) error {
 
 	r.clusterUUID = utils.GetClusterUUID(ctx, r.client)
 
+	r.log.Debugf("controller is ready")
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/source/controller.go:155]:
```suggestion
	r.log.Debug("controller is ready")
```

File: deckhouse-controller/pkg/controller/module-controllers/utils/utils.go [hunk 6]
```diff
@@ -220,3 +232,197 @@ func GetClusterUUID(ctx context.Context, cli client.Client) string {
 	// generate a random UUID if the key is missing
 	return uuid.Must(uuid.NewV4()).String()
 }
+
+func ValidateModule(def moduletypes.Definition, values addonutils.Values, logger *log.Logger) error {
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/utils/utils.go:236]: Proposal: 
Add this method to moduletypes.Definition struct

File: deckhouse-controller/pkg/controller/module-controllers/utils/utils.go [hunk 7]
```diff
@@ -220,3 +232,197 @@ func GetClusterUUID(ctx context.Context, cli client.Client) string {
...
+
+	if result != nil {
+		return fmt.Errorf("validate module: %w", result)
+	}
+
+	return nil
+}
+
+// EnableModule deletes old symlinks and creates a new one
+func EnableModule(downloadedModulesDir, oldSymlinkPath, newSymlinkPath, modulePath string) error {
```
[the remark this answers is not in the document]
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/module-controllers/utils/utils.go:283]: And injector too

[the remark this answers is not in the document]
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/module-controllers/utils/utils.go:283]: Refactoring in next pr

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @ipaqsa. I've started the workflow for label 'e2e/run/gcp'...

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 8]
```diff
@@ -64,309 +55,332 @@ import (
...
+func (r *reconciler) preflight(ctx context.Context) error {
+	defer r.init.Done()
+
+	// wait until module manager init
+	r.log.Debug("wait until module manager is inited")
+	if err := wait.PollUntilContextCancel(ctx, 2*time.Second, true, func(_ context.Context) (bool, error) {
+		return r.moduleManager.AreModulesInited(), nil
+	}); err != nil {
+		r.log.Errorf("failed to init module manager: %v", err)
+		return err
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:151]:
```suggestion
		return fmt.Errorf("init module manager: %w",err)
```

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go [hunk 9]
```diff
@@ -150,84 +132,84 @@ func (k *kubeAPI) PatchReleaseApplyAfter(release *v1alpha1.ModuleRelease, applyT
...
-	if err != nil {
+
+	if err = def.Validate(values, k.log); err != nil {
 		release.Status.Phase = v1alpha1.ModuleReleasePhaseSuspended
 		if statusErr := k.UpdateReleaseStatus(ctx, release, "validation failed: "+err.Error(), release.Status.Phase); statusErr != nil {
-			k.logger.Errorf("update the '%s' release status: %s", release.Name, statusErr.Error())
+			k.log.Errorf("update the '%s' release status: %v", release.Name, statusErr)
 		}
-		return fmt.Errorf("module '%s:v%s' validation failed: %s", moduleName, release.Spec.Version.String(), err)
+		return fmt.Errorf("the '%s:v%s' module validation: %w", release.Spec.ModuleName, release.Spec.Version.String(), err)
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/release/updater.go:181]: more clear
```suggestion
		return fmt.Errorf("the '%s: v%s' module validation: %w", release.Spec.ModuleName, release.Spec.Version.String(), err)
```

File: deckhouse-controller/pkg/controller/module-controllers/utils/utils.go [hunk 10]
```diff
@@ -220,3 +233,175 @@ func GetClusterUUID(ctx context.Context, cli client.Client) string {
...
+		if err = os.Remove(newSymlinkPath); err != nil {
+			return fmt.Errorf("delete the '%s' new symlink: %w", newSymlinkPath, err)
+		}
+	}
+
+	// make absolute path for versioned module
+	moduleAbsPath := filepath.Join(downloadedModulesDir, strings.TrimPrefix(modulePath, "../"))
+	// check that module exists on a disk
+	if _, err := os.Stat(moduleAbsPath); os.IsNotExist(err) {
+		return fmt.Errorf("the '%s' module absolute path not found", moduleAbsPath)
```
Comment (ldmonster) [inline deckhouse-controller/pkg/controller/module-controllers/utils/utils.go:259]: looks like place for a constant Err

File: deckhouse-controller/pkg/helpers/reginjector/injector.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+	return buf.Bytes(), err
+}
+
+// part of openapi schema for injecting registry values
+type registrySchemaForValues struct {
+	Type       string   `yaml:"type"`
+	Default    struct{} `yaml:"default"`
+	Properties struct {
+		Base struct {
```
Comment (ldmonster) [inline deckhouse-controller/pkg/helpers/reginjector/injector.go:80]: looks like we need another struct to make this more clear

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go (hunk 9 above)
Comment (ipaqsa) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go:181]: "test: v1" I think it is not more clear

File: deckhouse-controller/pkg/controller/module-controllers/utils/utils.go (hunk 10 above)
Comment (ipaqsa) [reply deckhouse-controller/pkg/controller/module-controllers/utils/utils.go:259]: Why ?

File: deckhouse-controller/pkg/helpers/reginjector/injector.go (hunk 11 above)
Comment (ipaqsa) [reply deckhouse-controller/pkg/helpers/reginjector/injector.go:80]: Agree, but it is not my code. Maybe in the next PR

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go (hunk 9 above)
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go:181]: "test:v1" vs "test: v1".
splitted information by spaces read's easier than without
it make sense if you want use ':' like separator after in code

File: deckhouse-controller/pkg/controller/module-controllers/utils/utils.go (hunk 10 above)
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/module-controllers/utils/utils.go:259]: i thinking about handling this error in upper layer. not necessary

File: deckhouse-controller/pkg/controller/module-controllers/release/updater.go (hunk 9 above)
Comment (ldmonster) [reply deckhouse-controller/pkg/controller/module-controllers/release/updater.go:181]: looks like you use docker like formatting.
it's ok

**Review by ldmonster — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/aws'...

**Review by yalosev — APPROVED** [review]

**Pull request merged**